### ISSUE TITLE
Training 2 distinct models to map different Inputs to Shared Vector Space

### ISSUE THREAD

**Issue by pertschuk** [UKPLab/sentence-transformers#29]
Hi - 
As I understand it, you choose to utilize the same model to encode both sentences, in lets say the NLI task. From cosine loss impl:
```python
def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
  reps = [self.model(sentence_feature)['sentence_embedding'] for sentence_feature in sentence_features]
```

In this case, since the first sentence represents a judgement and the later a hypothesis, would it make sense to train two separate models simultaneously to map onto the same vector space? 

Is this already what's happening and I am confused? Or is this something you've explored already.

For example something like: 
```python
def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
  rep_a = self.model_a(sentence_features[0])['sentence_embedding']
  rep_b = self.model_b(sentence_features[1])['sentence_embedding']
  output = torch.cosine_similarity(rep_a, rep_b)

```

Thank you
Jack

**Comment by nreimers** [comment]
Hi Jack,
If you have two quite different inputs, like question and answer, it can make sense to train two models. In that case, you get a function that is no longer symmetric, i.e. F(a, b)!= F(b, a)

What people also have done is to use the same base model (like Bert) for both inputs and then only add a single custom feed forward layer to one sentence. The outputs are then:
o_a = BERT(a) 
o_b = FeedForward(BERT(b)) 

This is currently not implemented in this repository. 

Best regards 
Nils Reimers

**Comment by pertschuk** [comment]
I'll leave this here in case anyone else tries this. I haven't tried adding the feedforward layer, but I created a class that takes 2 SentenceTransformers in its constructor and a loss function that calculates the cosine and fits based on that. 

I observed a slight boost in performance from just using sentence-transformers out of the box , albeit not statistically significant (improvement of .01 MRR in reranking claim / evidence pairs, still somewhat below feeding them in simultaneously to a model). Additionally it seems to take quite a lot longer to train (at least 30% slower per step) and requires more steps to fit. 

```python
    model_a = SentenceTransformer('bert-large-nli-mean-tokens')
    model_b = SentenceTransformer('bert-large-nli-mean-tokens')
    model = BiSentenceTransformer(model_a, model_b)
    train_batch_size = 2
    num_epochs = 1
    warmup_steps = 100
    model_save_path = './fever-output'
    reader = FEVERReader()
    train_examples = reader.get_examples('train',table='test.scorer_train_title')
    train_data = SentencesDataset(train_examples, model_a)
    train_dataloader = DataLoader(train_data, shuffle=True, batch_size=train_batch_size)
    train_loss = losses.BiCosineSimilarityLoss(model)

    dev_examples = reader.get_examples('dev',table='test.scorer_test_title')
    dev_data = SentencesDataset(examples=dev_examples, model=model_a)
    dev_dataloader = DataLoader(dev_data, shuffle=False, batch_size=train_batch_size)
    evaluator = EmbeddingSimilarityEvaluator(dev_dataloader)

    model.fit(train_objective=(train_dataloader, train_loss),
              evaluator=evaluator,
              epochs=num_epochs,
              evaluation_steps=20000,
              warmup_steps=warmup_steps,
              output_path_base=model_save_path)
```

**Comment by nreimers** [comment]
Thanks for posting the example here.

**Comment by Timoeller** [comment]
Hey @pertschuk nice work! We are also interested in using two separate encoders. Could you post the code for BiSentenceTransformer as well, please? (if you still have it)
Thanks
